### ISSUE TITLE
TraCI protocol enhancements

### ISSUE THREAD

**Issue by behrisch** [eclipse/sumo#8076]
Just a collection of ideas to be discussed

- [ ] remove the wrapping envelope which allows to send multiple traci messages at once (this only works in theory, there are no tests and the usefulness is questionable because usually you want the result immediately)
- [ ] send the traci version with every command (easier change of semantics without tinkering with compound sizes or new variable names; could replace the bits when removing the envelope (see above) maintaining some kind of compatibility)
- [ ] move everything which is not directly related to establishing and maintaining the connection to the simulation domain (at least `step` and `load`, TBD: complete the list of affected commands)
- [ ] always use signed bytes especially for bools and lane index / offsets (TBD: a list of affected commands)
- [ ] use the protocol consistently. Every command should be either a setter or a getter, every setter has parameters which may be a compound of zero length (TBD: a list of affected commands)

**Comment by sommer** [comment]
I don't know if this has been discussed and discarded yet, but it might be a good idea to move to a fully-featured third-party library for serialization/deserialization instead of iteratively improving TraCI's serialization capabilities. One that currently enjoys pretty widespread adoption are Google Protocol Buffers <https://github.com/protocolbuffers/protobuf>.

This would instantly make TraCI faster, more space/memory efficient, and more robust. Last but not least it might ease debugging.

Similarly, but one step further, we might consider moving from handling Remote Procedure Calls (RPC) using TraCI to a fully-featured third-party library. Here, I can't really offer any particular recommendation, but something like gRPC <https://github.com/grpc/grpc> might be worth looking into.

**Comment by dbuse** [comment]
I agree that using a full-fledged serialization library or even an RPC tool kit may be the better idea. This would alleviate most of the points described by @behrisch in the opening post. I see three major points here:
* Versioning -- will be built-in, e.g., by how Google Protocol Buffers does this with IDs and Extensions
* Performance -- probably better compression and highly-optimized (de)serialization, with compiled binaries for interpreted languages like Python
* Adaptation -- implementing TraCI (or even a portion of it) in another language will get much easier

There are multiple options, depending on where the focus lies. Google Protocol Buffers, Flatbuffers, Cap'n Proto, binary JSON, msgpack, ASN1, etc. So far I have made good experiences with Protocol Buffers (even in software that forwards data originally received from Sumo via TraCI), but would probably go for Flatbuffers for something new.

**Comment by allenwoods** [comment]
Hi, I'd like to know if there's been any update on this?

**Comment by behrisch** [comment]
Not really. We still want to switch to a new framework (probably protobuf or something equivalent) eventually but there is no current project demanding it.

**Comment by allenwoods** [comment]
I see. Is there a single entry point we can modify to support a new RPC tool, or do we have to modify multiple places?

**Comment by behrisch** [comment]
You can have a look at how libsumo or libtraci are implemented. We probably need something to wrap the whole (public) interface of libsumo, that is all header files in src/libsumo without the Helpers. If you are not choosing something too obscure we might as well do this together.
